prow/pod-utils/gcs/upload.go

```diff
@@ -65,14 +65,57 @@ func Upload(ctx context.Context, bucket, gcsCredentialsFile, s3CredentialsFile s
 		return fmt.Errorf("new opener: %w", err)
 	}
 	dtw := func(dest string) dataWriter {
-		compressFileTypeSet := sets.NewString(compressFileTypes...)
-		ext := strings.TrimPrefix(filepath.Ext(dest), ".")
-		compress := compressFileTypeSet.Has("*") || compressFileTypeSet.Has(ext)
+		compress := shouldCompressFile(dest, sets.New[string](compressFileTypes...))
 		return &openerObjectWriter{Opener: opener, Context: ctx, Bucket: parsedBucket.String(), Dest: dest, compress: compress}
 	}
 	return upload(dtw, uploadTargets)
 }
 
+func shouldCompressFile(dest string, compressFileTypes sets.Set[string]) bool {
+	ext := strings.TrimPrefix(filepath.Ext(dest), ".")
+	if ext == "gz" || ext == "gzip" {
+		return false
+	}
+	if !compressFileTypes.Has("*") && !compressFileTypes.Has(ext) {
+		return false
+	}
+
+	fileInfo, err := os.Stat(dest)
+	if err != nil {
+		logrus.WithError(err).Warnf("unable to stat file: %s to determine if we should compress", dest)
+		return false
+	}
+
+	// Don't compress files smaller than 1KB as it will likely result in a larger file
+	if fileInfo.Size() < 1024 {
+		return false
+	}
+
+	file, err := os.Open(dest)
+	if err != nil {
+		logrus.WithError(err).Warnf("unable to open file: %s to determine if we should compress", dest)
+		return false
+	}
+	defer file.Close()
+
+	header := make([]byte, 2)
```

Comment:
Let's use `http.DetectContentType()` so we don't have to write and own this logic.